src/local_executor.rs

```diff
@@ -0,0 +1,619 @@
+        // `LocalExecutor` is !Send, there is no way to have concurrent access to the
+        // values in `State`, including the active field.
+        let active = unsafe { &mut *self.state().active.get() };
+        Self::spawn_inner(state, future, active)
+    }
+
+    /// Spawns many tasks onto the executor.
+    ///
+    /// As opposed to the [`spawn`] method, this locks the executor's inner task lock once and
+    /// spawns all of the tasks in one go. With large amounts of tasks this can improve
+    /// contention.
+    ///
+    /// It is assumed that the iterator provided does not block; blocking iterators can lock up
+    /// the internal mutex and therefore the entire executor. Unlike [`Executor::spawn`], the
+    /// mutex is not released, as there are no other threads that can poll this executor.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// use async_executor::LocalExecutor;
+    /// use futures_lite::{stream, prelude::*};
+    /// use std::future::ready;
+    ///
+    /// # futures_lite::future::block_on(async {
+    /// let mut ex = LocalExecutor::new();
+    ///
+    /// let futures = [
+    ///     ready(1),
+    ///     ready(2),
+    ///     ready(3)
+    /// ];
+    ///
+    /// // Spawn all of the futures onto the executor at once.
+    /// let mut tasks = vec![];
+    /// ex.spawn_many(futures, &mut tasks);
+    ///
+    /// // Await all of them.
+    /// let results = ex.run(async move {
+    ///     stream::iter(tasks).then(|x| x).collect::<Vec<_>>().await
+    /// }).await;
+    /// assert_eq!(results, [1, 2, 3]);
+    /// # });
+    /// ```
+    ///
+    /// [`spawn`]: LocalExecutor::spawn
+    /// [`Executor::spawn_many`]: Executor::spawn_many
+    pub fn spawn_many<T: 'a, F: Future<Output = T> + 'a>(
+        &self,
+        futures: impl IntoIterator<Item = F>,
+        handles: &mut impl Extend<Task<F::Output>>,
+    ) {
+        let tasks = {
+            let state = self.state();
+
+            // SAFETY: All UnsafeCell accesses to active are tightly scoped, and because
+            // `LocalExecutor` is !Send, there is no way to have concurrent access to the
+            // values in `State`, including the active field.
+            let active = unsafe { &mut *state.active.get() };
+
+            // Convert the futures into tasks.
+            futures
+                .into_iter()
+                .map(move |future| Self::spawn_inner(state, future, active))
+        };
+
+        // Push the tasks to the user's collection.
+        handles.extend(tasks);
+    }
+
+    /// Spawn a future while holding the inner lock.
+    fn spawn_inner<T: 'a>(
+        state: Pin<&'a State>,
+        future: impl Future<Output = T> + 'a,
+        active: &mut Slab<Waker>,
+    ) -> Task<T> {
+        // Remove the task from the set of active tasks when the future finishes.
+        let entry = active.vacant_entry();
+        let index = entry.key();
+        let future = AsyncCallOnDrop::new(future, move || {
```

Comment:
It would be nice if `Builder::new().propagate_panic(true)` could be saved in a binding above this line.

It is not really necessary, but it makes it obvious that no panics can occur between the construction of `AsyncCallOnDrop` and the call to `.spawn_unchecked`, at which point the handling of drop unwinding gets taken over by `Task`, which automatically schedules itself in that case and notifies this `LocalExecutor`.

Note that `entry.insert` might allocate, and subsequently panic on OOM.
I think the easiest way to prevent panics in this function in a sort-of "critical section" w.r.t. `state.active` would be to invoke `active.reserve(1);` at the top of the function.